modules/cnvkit/1.0/cnvkit.smk

```diff
@@ -0,0 +1,318 @@
+#!/usr/bin/env snakemake
+
+
+##### ATTRIBUTION #####
+
+
+# Original Author:  N/A
+# Module Author:    Jasper Wong
+# Contributors:     N/A
+
+
+##### SETUP #####
+
+
+# Import package with useful functions for developing analysis modules
+import oncopipe as op
+
+# Setup module and store module-specific configuration in `CFG`
+# `CFG` is a shortcut to `config["lcr-modules"]["cnvkit"]`
+CFG = op.setup_module(
+    name = "cnvkit",
+    version = "1.0",
+    subdirectories = ["inputs", "cnvkit", "cn_calls", "SNPs", "BAF", "scatter", "outputs"],
+)
+
+# Define rules to be run locally when using a compute cluster
+localrules:
+    _cnvkit_input_bam,
+    _cnvkit_symlink_cns,
+    _cnvkit_output,
+    _cnvkit_all,
+
+
+##### RULES #####
+if CFG["options"]["hard_masked"] == True:
+    CFG["runs"]["masked"] = "_masked"
+else:
+    CFG["runs"]["masked"] = ""
+    
+wildcard_constraints:
+    masked = ".{0}|_masked",
+    genome_build = ".+(?<!masked)"
+    
+    
+# cnvkit reference files
+# gene annotation files
+rule _get_refFlat:
+    output:
+        refFlat = CFG["dirs"]["inputs"] + "ref/refFlat.txt"
+    params:
+        url = lambda w: {
+            "grch37": "http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/refFlat.txt.gz",
+            "hg19": "http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/refFlat.txt.gz",
+            "hs37d5": "http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/refFlat.txt.gz",
+            "grch38": "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/refFlat.txt.gz",
+            "hg38": "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/refFlat.txt.gz"
+        }
+    shell:
+        """
+            wget -O {output.refFlat} {params.url}
+        """
+
+
+# Symlinks the input files into the module results directory (under '00-inputs/')
+rule _cnvkit_input_bam:
+    input:
+        bam = CFG["inputs"]["sample_bam"],
+        bai = CFG["inputs"]["sample_bai"]
+    output:
+        bam = CFG["dirs"]["inputs"] + "{seq_type}--{genome_build}/{capture_space}/{sample_id}.bam",
+        bai = CFG["dirs"]["inputs"] + "{seq_type}--{genome_build}/{capture_space}/{sample_id}.bam.bai",
+        crai = CFG["dirs"]["inputs"] + "{seq_type}--{genome_build}/{capture_space}/{sample_id}.bam.crai"
+    run:
+        op.relative_symlink(input.bam, output.bam)
+        op.relative_symlink(input.bai, output.bai)
+        op.relative_symlink(input.bai, output.crai)
+
+
+def get_normals(wildcards):
+    CFG = config["lcr-modules"]["cnvkit"]
+    capture_space = CFG["runs"][CFG["runs"]["normal_capture_space"].isin([wildcards.capture_space])]
```

Comment:
Yes, it is to gather all normals that share the same capture space. CNVkit will run in batches based on a panel of normals you provide and a series of samples that would use the normals. So to separate analyses based on capture space, I have it as a wildcard. If capture_space is not a good column to use in the samples table, is there a better alternative to find samples that share the same capture kit?

Reply:
What does it do differently for the different capture space? Say, if there is a target panel of 100 genes - what will happen with the cnvkit results? I am thinking if we can apply the genome_build here instead. You can also take a look at how the lofreq is now implemented: it uses a switch on wildcard and a default bed/genome guild. If no bed file is provided, it will use the default bed - and if there is a bed file given, lofreq will use that. Maybe, we can implement something similar here? That function returns the path to bed file, but can give an idea how to get the list of samples
https://github.com/LCR-BCCRC/lcr-modules/blob/82439d6721d0b1587cd16200d19a7bfc1f4a27e8/modules/lofreq/1.1/lofreq.smk#L61-L81